benches/benches/agui_core/tree.rs

```diff
@@ -1,53 +1,56 @@
 use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use std::rc::Rc;
+use std::cell::RefCell;
 
 use agui::{element::ElementId, util::tree::Tree};
 
+// add and remove items from tree
 fn tree_ops(c: &mut Criterion) {
```

Comment:
// add and remove items from tree
fn tree_ops(c: &mut Criterion) {
    // add items to tree
    c.bench_function("add to tree", |b| {
        b.iter_with_setup(
            || {
                // setup the tree
                let mut tree = Tree::<ElementId, usize>::default();
               // add items to the tree
                let root_id = tree.add(None, 0);
                let mut widget_ids = Vec::with_capacity(1000);
                (tree, root_id, widget_ids)
            },
            |(mut tree, root_id, mut widget_ids)| {
                for i in 0..1000 {
                    widget_ids.push(tree.add(Some(root_id), i));
                }
                black_box(widget_ids);
            },
        )
    });

    // remove items from tree
    c.bench_function("remove from tree", |b| {
        b.iter_with_setup(
            || {
                // setup the tree
                let mut tree = Tree::<ElementId, usize>::default();
                let root_id = tree.add(None, 0);
                let mut widget_ids = Vec::with_capacity(1001);
                widget_ids.push(root_id);
                for i in 0..1000 {
                    widget_ids.push(tree.add(Some(root_id), i));
                }
                (tree, widget_ids)
            },
            |(mut tree, widget_ids)| {
                for widget_id in widget_ids.into_iter() {
                    black_box(tree.remove_item(widget_id));
                }
            },
        )
    });
}

Reply:
Added the use of `b.iter_with_setup` for `tree_ops` instead of `b.iter`. Preallocating memory and `remove_item` also incorporated.